File: app/src/test/java/com/veyndan/paper/reddit/NodeTest.java
package com.veyndan.paper.reddit;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.veyndan.paper.reddit.util.Node;

import org.junit.Test;

import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * To work on unit tests, switch the Test Artifact in the Build Variants view.
 */
public class NodeTest {

    @Test
    public void descendantCount_unknownDescendantCount_isCorrect() {
        final Node<Void> node30 = setAsChildren();

        final Node<Void> node20 = setAsChildren(node30);
        final Node<Void> node21 = setAsChildren();
        final Node<Void> node22 = setAsChildren();

        final Node<Void> node10 = setAsChildren();
        final Node<Void> node11 = setAsChildren(node20, node21);
        final Node<Void> node12 = setAsChildren(node22);

        final Node<Void> tree = setAsChildren(node10, node11, node12);

        tree.preOrderTraverse(0)
                .flatMapSingle(Node::descendantCount)
                .test()
                .assertValues(7, 0, 3, 1, 0, 0, 1, 0)
                .assertComplete();
    }

    @Test
    public void descendantCount_knownDescendantCount_isCorrect() {
        final Node<Void> node30 = setAsChildren(0);

        final Node<Void> node20 = setAsChildren(1, node30);
        final Node<Void> node21 = setAsChildren(0);
        final Node<Void> node22 = setAsChildren(0);

        final Node<Void> node10 = setAsChildren(0);
        final Node<Void> node11 = setAsChildren(3, node20, node21);
        final Node<Void> node12 = setAsChildren(1, node22);

        final Node<Void> tree = setAsChildren(7, node10, node11, node12);

        tree.preOrderTraverse(0)
                .flatMapSingle(Node::descendantCount)
                .test()
                .assertValues(7, 0, 3, 1, 0, 0, 1, 0)
                .assertComplete();
    }

    @Test
    public void depth_isCorrect() {
        final Node<Void> node30 = setAsChildren();

        final Node<Void> node20 = setAsChildren(node30);
        final Node<Void> node21 = setAsChildren();
        final Node<Void> node22 = setAsChildren();

        final Node<Void> node10 = setAsChildren();
        final Node<Void> node11 = setAsChildren(node20, node21);
        final Node<Void> node12 = setAsChildren(node22);

        final Node<Void> tree = setAsChildren(node10, node11, node12);

        tree.preOrderTraverse(0)
                .map(Node::depth)
                .test()
                .assertValues(0, 1, 1, 2, 3, 2, 1, 2)
                .assertComplete();
    }

    private static <T> Node<T> setAsChildren(final int descendantCount, final Node<T>... children) {
        return new Node<T>() {

            @Nullable
            @Override
            public Integer degree() {
                return null;
            }

            @NonNull
            @Override
            public Observable<Node<T>> children() {
                return Observable.fromArray(children);
            }

            @NonNull
            @Override
            public Single<Integer> descendantCount() {
                return Single.just(descendantCount);
            }
        };
    }

    private static <T> Node<T> setAsChildren(final Node<T>... children) {
        return new Node<T>() {

            @Nullable
            @Override
            public Integer degree() {
                return null;
            }

            @NonNull
            @Override
            public Observable<Node<T>> children() {
                return Observable.fromArray(children);
            }
        };
    }
}

File: app/src/test/java/com/veyndan/paper/reddit/NodeTest.kt
package com.veyndan.paper.reddit

import com.veyndan.paper.reddit.util.Node
import io.reactivex.Observable
import io.reactivex.Single
import org.junit.Test

/**
 * To work on unit tests, switch the Test Artifact in the Build Variants view.
 */
class NodeTest {

    @Test
    fun descendantCount_unknownDescendantCount_isCorrect() {
        val node30 = setAsChildren<Void>()

        val node20 = setAsChildren(node30)
        val node21 = setAsChildren<Void>()
        val node22 = setAsChildren<Void>()

        val node10 = setAsChildren<Void>()
        val node11 = setAsChildren(node20, node21)
        val node12 = setAsChildren(node22)

        val tree = setAsChildren(node10, node11, node12)

        tree.preOrderTraverse(0)
                .flatMapSingle<Int> { it.descendantCount() }
                .test()
                .assertValues(7, 0, 3, 1, 0, 0, 1, 0)
                .assertComplete()
    }

    @Test
    fun descendantCount_knownDescendantCount_isCorrect() {
        val node30 = setAsChildren<Void>(0)

        val node20 = setAsChildren(1, node30)
        val node21 = setAsChildren<Void>(0)
        val node22 = setAsChildren<Void>(0)

        val node10 = setAsChildren<Void>(0)
        val node11 = setAsChildren(3, node20, node21)
        val node12 = setAsChildren(1, node22)

        val tree = setAsChildren(7, node10, node11, node12)

        tree.preOrderTraverse(0)
                .flatMapSingle<Int> { it.descendantCount() }
                .test()
                .assertValues(7, 0, 3, 1, 0, 0, 1, 0)
                .assertComplete()
    }

    @Test
    fun depth_isCorrect() {
        val node30 = setAsChildren<Void>()

        val node20 = setAsChildren(node30)
        val node21 = setAsChildren<Void>()
        val node22 = setAsChildren<Void>()

        val node10 = setAsChildren<Void>()
        val node11 = setAsChildren(node20, node21)
        val node12 = setAsChildren(node22)

        val tree = setAsChildren(node10, node11, node12)

        tree.preOrderTraverse(0)
                .map { it.depth }
                .test()
                .assertValues(0, 1, 1, 2, 3, 2, 1, 2)
                .assertComplete()
    }

    private fun <T> setAsChildren(descendantCount: Int, vararg children: Node<T>): Node<T> {
        return object : Node<T>() {

            override fun children(): Observable<Node<T>> {
                return Observable.fromArray(*children)
            }

            override fun descendantCount(): Single<Int> {
                return Single.just(descendantCount)
            }
        }
    }

    private fun <T> setAsChildren(vararg children: Node<T>): Node<T> {
        return object : Node<T>() {

            override fun children(): Observable<Node<T>> {
                return Observable.fromArray(*children)
            }
        }
    }
}
